lib/server.go

```diff
@@ -930,15 +931,73 @@ func (srv *Server) GetBlocks(pp *Peer, maxHeight int) {
 		pp)
 }
 
+// shouldVerifySignatures determines if we should verify signatures for headers or not.
+// For PoW headers, this always returns true because there are no signatures to verify and there is
+// no impact on syncing.
+// For PoW blocks, we verify signatures if we're not syncing.
+// For PoS headers and blocks, we check if we've seen the checkpoint block.
+// If the checkpoint block info is nil, we return true so that we verify signatures.
+// If we haven't seen the checkpoint block yet, we skip signature verification.
+// If the header height does not match the checkpoint block height, we should disconnect the peer.
+// Otherwise, return true.
+func (srv *Server) shouldVerifySignatures(header *MsgDeSoHeader, isHeaderChain bool) (_verifySignatures bool, _shouldDisconnect bool) {
+	// For PoW headers, there is no signature to verify in the header, so we return true
+	// just to be safe, but it has no impact on the syncing.
+	// For PoW blocks, we verify signatures if we're not syncing.
+	if srv.params.IsPoWBlockHeight(header.Height) {
+		if !isHeaderChain {
+			return !srv.blockchain.isSyncing(), false
+		}
+		return true, false
+	}
+	// For PoS blocks, we check if we've seen the checkpoint block.
+	// If we don't have a check point block info, we return true so that we verify signatures.
+	if srv.blockchain.checkpointBlockInfo == nil {
+		return true, false
+	}
+	var hasSeenCheckpointBlockHash bool
+	var checkpointBlockNode *BlockNode
+	if isHeaderChain {
```

Comment:
I think there's a small bug here. If we NEVER find the checkpoint header then I think this code does not disconnect from the peer properly. Instead, we need to add as a condition that the height has not progressed beyond the checkpoint as well.

Reply:
I suggested the edit for this below